### ISSUE TITLE
Merge new upstream stubs in master

### ISSUE THREAD

**Issue by stlehmann** [stlehmann/PyQt5-stubs#42]
Thanks to @BryceBeagle we now have much more stubs to offer. I didn't have a closer look at them. But I guess they also need to be altered to be usable with mypy.

**Comment by stlehmann** [comment]
It might be a very pragmatic approach but for the new stubs I would maybe just take them and put them in the master branch, release them and fix errors on-the-go. Otherwise the work might be a bit overwhelming. @BryceBeagle, @The-Compiler what do you think?

**Comment by BryceBeagle** [comment]
I think the best long term solution is to have one branch. I've been thinking about a potential workflow for a while and would like to hear your opinions.

- The upstream stubs live on the master branch
- The changes we make don't actually get applied to the upstream stubs, but instead reside in a set of patch files
- The patch files could correspond to (and better document) the issues that currently exist in the [issues.md](https://github.com/stlehmann/PyQt5-stubs/blob/master/issues.md)
- When a new version of PyQt is released, just make sure the patch files still properly apply by fixing the (presumably) minor changes that occurred
- When we want to release a new version of the patched stubs, we can just apply the patches locally and then build/deploy a wheel from those files
- Tests could be written such that we test to make sure a mypy error occurs without the patches, and then does not occur after a patch has been applied

\* We might even not want to store the upstream stubs, but only store the patches. A repo initialization script could just build the stubs with the Dockerfile (or pull from the PyPI wheels when they contain the stubs) for development purposes.

**Comment by The-Compiler** [comment]
@stlehmann Fixing errors as they come up sounds okay - there are probably a lot of unfixed issues (like more signals) in the existing stubs anyways.

@BryceBeagle As long as we hand-edit the stubs (rather than automating the edits in some way), I'd rather keep the current workflow. Mixing patch-files and git just seems to complicate the workflow for everyone - it's much harder for contributors to edit, it's harder to review, and it's much harder to get the end result. I'd rather use git's functionality to get me diffs in whatever way seems to help, rather than storing diffs in a git repo.

**Comment by stlehmann** [comment]
I actually like the idea of @BryceBeagle to use patch files. At the moment I find it VERY cumbersome to go through all the code and see what has changed after each release. We actually have no overview which changes have been applied. But I agree with @The-Compiler that using git patch files would make things even more complicated because they are hard to maintain.
I would prefer a much clearer format that allows easy editing and provides a clear view what changes have been appied and where.

How about a JSON-file with the following format:

    {
        "file": "QtWidgets.pyi",
        "patches": [
            {
                "operation": "edit"  # can be edit or insert
                "line": 51,
                "org": "def windowIconTextChanged(self, iconText: str) -> None: ..."
                "new": "windowIconChanged: QtCore.pyqtSignal"
            },
        ]
    }

We always check if the given line contains the original content. This way we will recognize if content has been moved in the original stubs, e.g. if a new line has been inserted.

**Comment by The-Compiler** [comment]
> At the moment I find it VERY cumbersome to go through all the code and see what has changed after each release. We actually have no overview which changes have been applied.

Doesn't commands like `git diff upstream..master` or `git show upstream` etc. give you exactly that?

I don't see the benefits to write a custom patch format if it just replaces lines. If we're going to write a custom patcher, IMHO it should be more "semantic", i.e. only take a list of all signals as input and then edit the stubs on its own based on that (essentially option 2 in https://github.com/stlehmann/PyQt5-stubs/issues/6#issuecomment-617598596)

**Comment by stlehmann** [comment]
> I don't see the benefits to write a custom patch format if it just replaces lines.

True, if you phrase it like this it might not be worth the effort. Though I'm still not satisfied with the current workflow.

> take a list of all signals as input and then edit the stubs on its own based on that (essentially option 2 in #6 (comment))

I like the idea of that. That would take a huge effort away from us.

**Comment by stlehmann** [comment]
#44 merged the new stubs files including QtWebkit and Qt3D. Now our CI will of course throw endless errors as the new stubs are not edited, yet.

**Comment by The-Compiler** [comment]
I'll continue the automation discussion in #6.

**Comment by altendky** [comment]
What is the current workflow?  One branch for upstream and one branch for our changes?  When upstream updates, commit to the upstream branch then merge into our branch?  or...

**Comment by altendky** [comment]
I'm going through the merge of existing upstream to master and it's a mess.  The merge base has a bunch of the customizations in it.  I haven't dug into the history but I think we need to make sure this doesn't happen in the future.  Git needs to be able to see a clean diff between the merge base and upstream that contains only actual upstream changes.  Presently it sees that upstream changed our customizations and so it thinks upstream needs to bring a change into the merge which makes for piles of conflicts.  I believe the rule is 'never merge non-upstream into upstream'.  Upstream needs to stay pure.

(or maybe I still haven't figured out the workflow here)

**Comment by altendky** [comment]
So I think the auto-merge probably threw away any of our customizations that we hadn't changed 'recently' due to only upstream having 'changed' them.  It looks like maybe just https://github.com/stlehmann/PyQt5-stubs/commit/2ce7aa9f6b9a8825cab395dbbdf932c3b8a9334a is the problem so I'll try merging from right before that instead and see what happens.

**Comment by stlehmann** [comment]
> What is the current workflow? One branch for upstream and one branch for our changes? When upstream updates, commit to the upstream branch then merge into our branch? or...

@altendky Yes, basically the Upstream branch contains the built and unchanged stubs from the PyQt5 repository. If there are changes on the Upstream branch they need to be merged in the master branch.

> Upstream needs to stay pure.

Yes, that is basically the idea.

**Comment by stlehmann** [comment]
Are you refering to PR #64? It looks good for me on first glance. Where are you having trouble specifically?

**Comment by altendky** [comment]
No, those were straightforward.  I haven't pushed anything related to the troublesome merge.  Merging upstream into master (even without either 5.15) is a mess because this is actually the first time that has happened.  It's looking a lot cleaner now that I'm excluding the dummy merge of master into upstream, but still not great.  I think that actually the very first upstream commit should have been dummy merged back into master immediately so that git would realize it is included in master.  Then any future merge of upstream into master would exclusively merge actual upstream changes since the merge base would be on upstream, not master.  Trying that now...

**Comment by altendky** [comment]
Yeah, that was a way smoother merge, even though it was 'three'.  See #63 

1. Ignore merge original upstream to master
2. Real merge almost latest upstream to master
3. Ignore merge latest upstream to master (because it was just an ignore merge itself)

I'll go back and try to figure out the SIP stub build, then update the 5.15.0 and 5.15.1 branches accordingly.

**Comment by altendky** [comment]
Ok...  So without presuming I've done everything right, here's what I see as the groupings to go through my PRs.  Each group should be completed before working on the next.  Hopefully I'm going down the expected and desired path with this work.

1. Tool enhancements
   - [x] #62 --jobs and --no-cache
   - [x] #66 Get `sip.pyi` in the Docker build
2. Update upstream
   - [x] #65 Merge (existing) upstream into master
   - [x] #63 5.15.0
   - [x] #64 5.15.1
3. Update master
   - [x] #70 Merge 5.15.0 to master
   - [x] #71 Merge 5.15.1 to master
4. The other fixups
   - [ ] #56 bound/unbound signals
   - [ ] #58 `QObject.findChild`[`ren`]
   - [ ] #59 `QGuiApplication.lastWindowClosed`
   - [ ] #61 mypy error related cleanup

**Comment by stlehmann** [comment]
@altendky, @BryceBeagle As you are both highly engaged I invited both of you as collaborators of this project. I hope this is OK for you and adds to the progress of this project :)

**Comment by altendky** [comment]
@BryceBeagle, thanks for all the reviews.  I definitely owe you...  something.  Let me know if there's something else I can contribute reviews or code or whatever to.  (or pieces you would like to fix here and have me review...)

**Comment by BryceBeagle** [comment]
Reviewing is the easy part 👍

**Comment by altendky** [comment]
With #71 merged to master this should be done.  I don't know how releases have been managed but I expect to keep making progress on the other bits in the near future if that matters.
